game-core/src/main/java/games/strategy/engine/framework/startup/ui/PlayerTypes.java

```diff
@@ -0,0 +1,167 @@
+import games.strategy.triplea.ai.pro.ProAi;
+import games.strategy.triplea.ai.weak.DoesNothingAi;
+import games.strategy.triplea.ai.weak.WeakAi;
+import games.strategy.triplea.settings.ClientSetting;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Value;
+import org.triplea.injection.Injections;
+
+@Value
+public class PlayerTypes {
+
+  public static final Type HUMAN_PLAYER =
+      new Type("Human") {
+        @Override
+        public Player newPlayerWithName(final String name) {
+          return new TripleAPlayer(name) {
+            @Override
+            public Type getPlayerType() {
+              return HUMAN_PLAYER;
+            }
+          };
+        }
+      };
+  public static final Type WEAK_AI =
+      new Type("Easy (AI)") {
+        @Override
+        public Player newPlayerWithName(final String name) {
+          return new WeakAi(name);
+        }
+      };
+  public static final Type FAST_AI =
+      new Type("Fast (AI)") {
+        @Override
+        public Player newPlayerWithName(final String name) {
+          return new FastAi(name);
+        }
+      };
+  public static final Type PRO_AI =
+      new Type("Hard (AI)") {
+        @Override
+        public Player newPlayerWithName(final String name) {
+          return new ProAi(name);
+        }
+      };
+  public static final Type DOES_NOTHING_AI =
+      new Type("Does Nothing (AI)") {
+        @Override
+        public Player newPlayerWithName(final String name) {
+          return new DoesNothingAi(name);
+        }
+      };
+  /** A hidden player type to represent network connected players. */
+  public static final Type CLIENT_PLAYER =
+      new Type("Client", false) {
+        @Override
+        public Player newPlayerWithName(final String name) {
+          return new TripleAPlayer(name) {
+            @Override
+            public Type getPlayerType() {
+              return CLIENT_PLAYER;
+            }
+          };
+        }
+      };
+  /** A 'dummy' player type used for battle calc. */
+  public static final Type BATTLE_CALC_DUMMY =
+      new Type("None (AI)", false) {
+        @Override
+        public Player newPlayerWithName(final String name) {
+          throw new UnsupportedOperationException(
+              "Fail fast - bad configuration, should instantiate dummy player "
+                  + "type only for battle calc");
+        }
+      };
+
+  Collection<Type> playerTypes = new ArrayList<>();
+
+  public PlayerTypes() {
+    playerTypes.add(HUMAN_PLAYER);
+    playerTypes.add(WEAK_AI);
+    playerTypes.add(FAST_AI);
+    playerTypes.add(PRO_AI);
+    playerTypes.add(DOES_NOTHING_AI);
+    playerTypes.add(CLIENT_PLAYER);
+    playerTypes.add(BATTLE_CALC_DUMMY);
+
+    final Injections injection = Injections.getInstance();
```

Comment:
Injections is mean to be very much a roll-your-own dependency injection framework. A framework like Dagger or Spring or Guice were voted down in favor of not having one some long time ago, yet DI is useful specifically for being able to have top level projects inject configuration into lower level layers. That is the background for how it is intended to be used.

With that context, I think Injections should own the full playerTypes list. As-is, we have an existing list of player types and then we append to that list from injections, which spreads the ownership around. I would envision that injections would have exactly the above `PlayerTypes.add(..)` code and would then be able to include the flow-field AI. Then any usages of injections simply needs to call 'injections.getPlayerTypes()'.

Ideally we would avoid the static dependency and try to keep the imperative shell, any outer-layer code in control of injections and then follow DI to pass the player types object to the method or class that needs it when we instantiate it.

Does that make sense? Any thoughts to changing this towards that goal, any hidden problems with that?

Reply:
`Injections` now stores all of the player types.  It is still passing them into `PlayerTypes` when used.  Do you want me to change those to static methods?  I could do that as well.

Reply:
`It is still passing them into PlayerTypes when used.`

How do you mean? I think anything that uses `PlayerTypes` should have it constructor injected. This avoids any tests from having to be aware of `Injections` and avoids any client of injected values from knowing how the values are injected as well.

Reply:
`PlayerTypes` is now really a bunch of utility methods.  So it could be changed back to static methods.